### PR TITLE
M5396 Attribute nullable expressions

### PR DESCRIPTION
**Pull request by dennishendriksen**
Example file: [it_emx_nillableExpression.xlsx](https://github.com/molgenis/molgenis/files/1406242/it_emx_nillableExpression.xlsx)
Note: test plan doesn't require updating, but ```Chr6_vragen_emx_testimport10.xlsx``` referenced in the test plan needs to be updated as part of [M4799](http://www.molgenis.org/ticket/4799).

#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- [x] User documentation updated <- I assume we'll update emx documentation once client code has been implemented
- [x] (If you have changed REST API interface) view-swagger.ftl updated
- [x] Test plan template updated
- [x] Clean commits

### CODE REVIEW

File: molgenis-data/src/main/java/org/molgenis/data/meta/model/AttributeMetadata.java [hunk 1]
```diff
@@ -160,6 +165,7 @@ public void init()
 		addAttribute(IS_READ_ONLY).setDataType(BOOL).setNillable(false).setLabel("Read-only");
 		addAttribute(IS_UNIQUE).setDataType(BOOL).setNillable(false).setLabel("Unique");
 		addAttribute(TAGS).setDataType(MREF).setRefEntity(tagMetadata).setLabel("Tags");
+		addAttribute(NULLABLE_EXPRESSION).setDataType(SCRIPT).setNillable(true).setLabel("Nullable expression");
```
Comment (fdlk) [inline molgenis-data/src/main/java/org/molgenis/data/meta/model/AttributeMetadata.java]: breaking change -> upgrade version number to 6.0.0-SNAPSHOT

File: molgenis-data-validation/src/main/java/org/molgenis/data/validation/EntityAttributesValidator.java [hunk 2]
```diff
@@ -39,7 +40,8 @@ public EntityAttributesValidator(ExpressionValidator expressionValidator)
 
 	public Set<ConstraintViolation> validate(Entity entity, EntityType meta)
 	{
-		Set<ConstraintViolation> violations = checkValidationExpressions(entity, meta);
+		Set<ConstraintViolation> violations = checkNullableExpressions(entity, meta);
```
Comment (fdlk) [inline molgenis-data-validation/src/main/java/org/molgenis/data/validation/EntityAttributesValidator.java]: Filling in a nullable expression will currently break the usability of the forms (required indication+validation / required-only button). I assume we consider the EMX format change a toggle? "As long as you don't use this, the client won't break"?

File: molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java [hunk 3]
```diff
@@ -110,13 +110,15 @@ private EditorAttribute toEditorAttribute(Attribute attribute)
 		boolean readonly = attribute.isReadOnly();
 		boolean unique = attribute.isUnique();
 		ImmutableList<EditorTagIdentifier> tags = tagMapper.toEditorTags(attribute.getTags());
+		String nullableExpression = attribute.getNullableExpression();
```
[the remark this answers is not in the document]
Comment (fdlk) [reply molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java]: What changes are eventually allowed in the metadata editor (once client implemented)? If attribute is nullable and you want to add nullableExpression, there's no code yet to validate existing rows in database. So I think you'll either have to forbid this update or add checking code that goes off when the metadata is changed.

File: molgenis-data-rest/src/main/java/org/molgenis/data/rest/AttributeResponse.java [hunk 4]
```diff
@@ -207,6 +208,12 @@ public AttributeResponse(final String entityParentName, EntityType entityType, A
 		}
 		else this.visible = null;
 
+		if (attributesSet == null || attributesSet.contains("nullableExpression".toLowerCase()))
+		{
+			this.nullableExpression = attr.getNullableExpression();
+		}
+		else this.visibleExpression = null;
```
[the remark this answers is not in the document]
Comment (dennishendriksen) [reply molgenis-data-rest/src/main/java/org/molgenis/data/rest/AttributeResponse.java]: whoops! fixed.

File: molgenis-data-validation/src/main/java/org/molgenis/data/validation/EntityAttributesValidator.java (hunk 2 above)
Comment (dennishendriksen) [reply molgenis-data-validation/src/main/java/org/molgenis/data/validation/EntityAttributesValidator.java]: Without the front-end changes the form will only be validated server-side, not client-side. The front-end work is scheduled as part of the next release.

File: molgenis-data/src/main/java/org/molgenis/data/meta/model/AttributeMetadata.java (hunk 1 above)
Comment (dennishendriksen) [reply molgenis-data/src/main/java/org/molgenis/data/meta/model/AttributeMetadata.java]: lets do this in a separate pull request somewhere before the release.

File: molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dennishendriksen) [reply molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java]: Interesting ... for validation expressions we have the same issue. Updating one of these expressions should first check whether the expression is valid for existing data.

[the remark this answers is not in the document]
Comment (fdlk) [reply molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java]: So the plan is to leave this up to a future story that also handles this for validation expression updates?

File: molgenis-data-import/src/main/java/org/molgenis/data/importer/emx/EmxMetaDataParser.java [hunk 5]
```diff
@@ -838,7 +838,16 @@ private void parseAttributesSheet(Repository<Entity> attributesRepo, Intermediat
 			}
 
 			if (emxAttrNillable != null)
-				attr.setNillable(parseBoolean(emxAttrNillable, rowIndex, EMX_ATTRIBUTES_NILLABLE));
+			{
+				if (emxAttrNillable.equalsIgnoreCase("true") || emxAttrNillable.equalsIgnoreCase("false"))
+				{
+					attr.setNillable(parseBoolean(emxAttrNillable, rowIndex, EMX_ATTRIBUTES_VISIBLE));
```
Comment (fdlk) [inline molgenis-data-import/src/main/java/org/molgenis/data/importer/emx/EmxMetaDataParser.java]: typo

File: molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dennishendriksen) [reply molgenis-metadata-manager/src/main/java/org/molgenis/metadata/manager/mapper/AttributeMapper.java]: Pending @mswertz (see mail)

**Pull request merged**